src/optimizer/cse_optimizer.cpp

```diff
@@ -63,14 +63,35 @@ void CommonSubExpressionOptimizer::CountExpressions(Expression &expr, CSEReplace
 		auto node = state.expression_count.find(expr);
 		if (node == state.expression_count.end()) {
 			// first time we encounter this expression, insert this node with [count = 1]
-			state.expression_count[expr] = CSENode();
+			// but only if it is not an interior argument of a short circuit sensitive expression.
+			if (std::accumulate(state.short_circuits.begin(), state.short_circuits.end(), idx_t(0)) == 0) {
+				state.expression_count[expr] = CSENode();
+			}
 		} else {
 			// we encountered this expression before, increment the occurrence count
 			node->second.count++;
 		}
 	}
-	// recursively count the children
-	ExpressionIterator::EnumerateChildren(expr, [&](Expression &child) { CountExpressions(child, state); });
+
+	// If we have a function that uses short circuiting, then we can only extract CSEs from the leftmost
+	// side of the argument tree (child_no == 0)
+	switch (expr.GetExpressionClass()) {
+	case ExpressionClass::BOUND_CONJUNCTION:
+	case ExpressionClass::BOUND_CASE: {
+		// Add a short circuit reference
+		state.short_circuits.emplace_back(0);
```

Comment:
Tracking this using a `short_circuits` vector is a bit cryptic.

For improved readability purposes - perhaps we can add a third parameter go `CountExpressions` - ` bool can_perform_cse` - which defaults to true and is set to false starting at the second child of a case/conjunction expression?

Reply:
Yeah I was thinking in terms of recursive tree processing, but I think we could just use a Boolean instead of the array.